source/AttributeStore.h

```diff
@@ -0,0 +1,233 @@
+/* AttributeStore.h
+Copyright (c) 2023 by tibetiroka
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef ATTRIBUTE_STORE_H_
+#define ATTRIBUTE_STORE_H_
+
+#include "attribute/Attribute.h"
+#include "attribute/AttributeAccess.h"
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Dictionary.h"
+
+#include <cmath>
+#include <functional>
+#include <limits>
+#include <map>
+#include <string>
+
+class AttributeStore {
+public:
+	// Checks whether the specified attribute is defined here.
+	bool IsPresent(const AttributeAccess attribute) const;
+	template <class T>
+	bool IsPresent(const T &attribute) const;
+	bool IsPresent(const char *attribute) const;
+	// Gets the value of the specified attribute, or 0 if not present.
+	double Get(const AttributeAccess attribute) const;
+	template <class T>
+	double Get(const T &attribute) const;
```

Comment:
I feel that we should not use templated functions here. Templated functions are very nice for generic storage classes (like Vector), but attributes should have only a very limited set of ways for addressing/accessing them.

My expectation is that we can address attributes by: 
- A single string (`const char*` and/or `const string &`) based on the "classic/legacy" attribute names.
- Two strings (category and effectType), for categorized effects.
- An Accessor object (that contains a category and effectType), only for the categorized effects.

Just having explicit get and set functions for those types of accesses feels more appropriate than templated functions.

Reply:
We can also choose to only have an Accessor object for accessing attributes.

Reply:
An example of the two functions:

```
	/**
	 * @param category The category, if accessing a categorized effect. Nullptr otherwise.
	 * @param effectTypeOrName The effect type, or name of the attribute to access if not a categorized effect or legacy type access.
	 */
	double Get(const char* category, const char* effectTypeOrName)

	/**
	 * @param attributeXs The identifier of the attribute, categorized effects only.
	 */
	double Get(AttributeAccessor attributeXs)
```

Reply:
Categorized attributes really shouldn't get string accessors, but I can probably simplify the rest.

Currently, there is an accessor via `AttributeAccess`, one with `AnyAttribute`, one with `std::string` and one with `const char *`. I suppose those could be replaced with a single AnyAttribute accessor, since std::variant should be constructible from all of those types. I'll have to see if that actually works, though.